twilight-lavalink/src/model/incoming.rs

```diff
@@ -0,0 +1,363 @@
+#[non_exhaustive]
+#[serde(rename_all = "camelCase")]
+pub enum Severity {
+    /// The cause is known and expected, indicates that there is nothing wrong with the library itself.
+    Common,
+    /// The probable cause is an issue with the library or there is no way to tell what the cause might be. This is the default level and other levels are used in cases where the thrower has more in-depth knowledge about the error.
+    Fault,
+    /// The cause might not be exactly known, but is possibly caused by outside factors. For example when an outside service responds in a format that we do not expect.
+    Suspicious,
+}
+
+/// The exception with the details attached on what happened when making a query to lavalink.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[non_exhaustive]
+#[serde(rename_all = "camelCase")]
+pub struct Exception {
+    /// The cause of the exception.
+    pub cause: String,
+    /// The message of the exception.
+    pub message: Option<String>,
+    /// The severity of the exception.
+    pub severity: Severity,
+}
+
+/// An incoming event from a Lavalink node.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[non_exhaustive]
+#[serde(untagged)]
+pub enum IncomingEvent {
+    /// Dispatched when player or voice events occur.
+    Event(Event),
+    /// Dispatched when you successfully connect to the Lavalink node.
+    Ready(Ready),
+    /// New statistics about a node and its host.
+    Stats(Stats),
+    /// An update about the information of a player.
+    PlayerUpdate(PlayerUpdate),
+}
+
+impl From<Ready> for IncomingEvent {
+    fn from(event: Ready) -> IncomingEvent {
+        Self::Ready(event)
+    }
+}
+
+impl From<Event> for IncomingEvent {
+    fn from(event: Event) -> IncomingEvent {
+        Self::Event(event)
+    }
+}
+
+impl From<PlayerUpdate> for IncomingEvent {
+    fn from(event: PlayerUpdate) -> IncomingEvent {
+        Self::PlayerUpdate(event)
+    }
+}
+
+impl From<Stats> for IncomingEvent {
+    fn from(event: Stats) -> IncomingEvent {
+        Self::Stats(event)
+    }
+}
+
+/// The discord voice information that lavalink uses for connection and sending information.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[non_exhaustive]
+#[serde(rename_all = "camelCase")]
+pub struct VoiceState {
+    /// The Discord voice endpoint to connect to.
+    pub endpoint: String,
+    /// The Discord voice session id to authenticate with. Note this is separate from the lavalink session id.
+    pub session_id: String,
+    /// The Discord voice token to authenticate with.
+    pub token: String,
+}
+
+/// An update about the information of a player. Filters are currently unsupported
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[non_exhaustive]
+#[serde(rename_all = "camelCase")]
+pub struct PlayerUpdate {
+    /// The guild ID of the player.
+    pub guild_id: Id<GuildMarker>,
+    /// Op code for this websocket event.
+    pub op: Opcode,
```

Comment:
I believe this should be an associated constant rather than part of the deserialized payload. Also, you can tag the `IncomingEvent` enum based on this field, which is what you appear to desire.

Reply:
See https://serde.rs/enum-representations.html

Reply:
I'll fix this in a follow up PR

Reply:
I believe this should be done as part of this PR, don't think a follow-up would be merged any time soon seeing how we're progressing on this one...